core/shared/src/main/scala/scalaz/zio/stm/STM.scala

```diff
@@ -0,0 +1,584 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package scalaz.zio.stm
+
+import java.util.concurrent.atomic.{ AtomicBoolean, AtomicLong, AtomicReference }
+
+import scalaz.zio.{ IO, UIO }
+
+import scala.collection.mutable.{ Map => MutableMap }
+import scala.util.{ Failure, Success, Try }
+
+/**
+ * `STM[E, A]` represents an effect that can be performed transactionally,
+ * resulting in a failure `E` or a value `A`.
+ *
+ * {{{
+ * def transfer(receiver: TRef[Int],
+ *              sender: TRef[Int], much: Int): UIO[Int] =
+ *   STM.atomically {
```

Comment:
Since most of the time ppl will use `STM` using this idiom (`STM.atomically { … } `) maybe make `STM.apply` an alias?

Reply:
There is already the much sorter `.commit` method on STM. Also, to be consistent with other effect types in ZIO, `apply` would be reserved for `succeed` or `succeedLazy`.